### PR TITLE
 kvserver: only pause followers when holding active lease

### LINKED ISSUE #84884: kv: paused followers interacts poorly with leader-not-leaseholder state
In a cluster that was overloaded due to an index backfill, I saw a large number of ranges serving foreground traffic begin hitting circuit breaker errors. Digging in deeper, it became clear that all ranges in this state had split leaders and leaseholders, and the leaseholders were on an overloaded node. Earlier in the test, I had set `admission.kv.pause_replication_io_threshold` to 0.8 to avoid replicating to overloaded followers.

The combination of paused replicas and the leader-not-leaseholder split effectively caused unavailability. Even though the leaseholder could propose writes, it would never hear about their result, so it would never acknowledge the result of those writes to clients.

Should we allow the leaseholder to be paused?

----

On a related note, I noticed that while all nodes had a non-zero value for the `admission.raft.paused_replicas`, the overloaded node itself (purple) had a few spikes where it reported thousands of paused replicas. My understanding is that this metric is reported from leader side, not the follower side, so I don't understand this. Is there any reason why an overloaded node would start pausing other replicas?

Screen Shot 2022-07-21 at 5 51 46 PM

Also, the description of this metric says: `The count is emitted by the leaseholder of each range.`

Should this instead say: `The count is emitted by the leader of each range.`

Jira issue: CRDB-17924

Epic CRDB-15069

### PR DESCRIPTION
**Pull request by tbg**
If the raft leader is not the leaseholder (which includes the case in
which we just transferred the lease away), leave all followers unpaused.
Otherwise, the leaseholder won't learn that the entries it submitted
were committed which effectively causes range unavailability.

Fixes #84884.

Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_raft_overload.go [hunk 1]
```diff
@@ -206,3 +206,73 @@ func (osm *overloadedStoresMap) Swap(
...
+	if !quotaPoolEnabledForRange(*r.descRLocked()) {
+		// If the quota pool isn't enabled (like for the liveness range), play it
+		// safe. The range is unlikely to be a major contributor to any follower's
+		// I/O and wish to reduce the likelihood of a problem in replication pausing
+		// contributing to an outage of that critical range.
+		return
+	}
+
+	status := r.leaseStatusAtRLocked(ctx, r.Clock().NowAsClockTimestamp())
+	if !status.IsValid() || status.OwnedBy(r.StoreID()) {
```
Comment (tbg) [inline pkg/kv/kvserver/replica_raft_overload.go:234]: missing a `!` here, caught by the test

**Comment by tbg** [comment]
Note: also still need to make sure a leader never considers itself paused.

File: pkg/kv/kvserver/replica_raft_overload.go [hunk 2]
```diff
@@ -206,3 +206,74 @@ func (osm *overloadedStoresMap) Swap(
...
+	}
+
+	status := r.leaseStatusAtRLocked(ctx, r.Clock().NowAsClockTimestamp())
+	if !status.IsValid() || !status.OwnedBy(r.StoreID()) {
+		// If we're not the leaseholder (which includes the case in which we just
+		// transferred the lease away), leave all followers unpaused. Otherwise, the
+		// leaseholder won't learn that the entries it submitted were committed
+		// which effectively causes range unavailability.
+		return
+	}
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_raft_overload.go:240]: Unpausing if there's no valid lease seems like a prudent failsafe.

However, it seems a bit lenient to unpause all followers when we don't have the lease. There might be cases where this situation persists, or where it causes intermittent unpausing. Should we only exclude the leaseholder instead, or are we worried that might open for other pathologies?

Comment (tbg) [reply pkg/kv/kvserver/replica_raft_overload.go:240]: Really hard to say without the benefit of hindsight. But I'd much rather approach pausing from below (adding complexity only when proven relevant) vs trying to anticipate problems and then (perhaps painfully) finding out that there was something we missed.
I expect it to be easier to see that a node is still sending to an overloaded follower (esp. in light of the new admission control logs that are landing ~right about now) vs figuring out that some loss of quorum or other degradation is caused by pausing.

**Review by erikgrinaker — APPROVED** [review]

**Comment by tbg** [comment]
bors r=erikgrinaker
TFTR!

**Pull request merged**